File: src/ruby/tormenta20/models/divindade.rb
# frozen_string_literal: true

require "active_record"
require "json"

module Tormenta20
  module Models
    # Model representing a Divindade (Deity) in Tormenta20.
    #
    # Deities grant powers to their followers and have specific energy types.
    #
    # @example Find deities with positive energy
    #   Divindade.energia_positiva
    #
    # @example Get deity's granted powers
    #   khalmyr = Divindade.find("khalmyr")
    #   khalmyr.granted_powers  # => ["coragem_total", "espada_justiceira", ...]
    #
    # @!attribute [rw] id
    #   @return [String] Unique identifier (e.g., "khalmyr")
    # @!attribute [rw] name
    #   @return [String] Deity name (e.g., "Khalmyr")
    # @!attribute [rw] title
    #   @return [String] Deity title (e.g., "O Deus da Justiça")
    # @!attribute [rw] energy
    #   @return [String] Energy type: "positiva", "negativa", or "qualquer"
    # @!attribute [rw] granted_powers
    #   @return [Array<String>] List of power IDs granted to followers
    class Divindade < Base
      self.table_name = "divindades"

      # Valid energy types for deities.
      ENERGIES = %w[positiva negativa qualquer].freeze

      validates :id, presence: true, uniqueness: true
      validates :name, presence: true
      validates :energy, presence: true, inclusion: { in: ENERGIES }

      # @!group Scopes

      # @!method energia_positiva
      #   Filter deities with positive energy.
      #   @return [ActiveRecord::Relation<Divindade>]
      scope :energia_positiva, -> { where(energy: "positiva") }

      # @!method energia_negativa
      #   Filter deities with negative energy.
      #   @return [ActiveRecord::Relation<Divindade>]
      scope :energia_negativa, -> { where(energy: "negativa") }

      # @!method energia_qualquer
      #   Filter deities with neutral/any energy.
      #   @return [ActiveRecord::Relation<Divindade>]
      scope :energia_qualquer, -> { where(energy: "qualquer") }

      # @!method by_energy(energy)
      #   Filter deities by energy type.
      #   @param energy [String] Energy type (positiva, negativa, qualquer)
      #   @return [ActiveRecord::Relation<Divindade>]
      scope :by_energy, ->(e) { where(energy: e) }

      # @!endgroup

      # Get races that typically worship this deity.
      #
      # @return [Array<String>] List of race names
      def races
        devotees&.dig("races") || []
      end

      # Get classes that typically worship this deity.
      #
      # @return [Array<String>] List of class names
      def classes
        devotees&.dig("classes") || []
      end

      # Convert deity to a hash representation.
      #
      # @return [Hash] Hash with all deity attributes
      def to_h
        {
          id: id,
          name: name,
          title: title,
          description: description,
          beliefs_objectives: beliefs_objectives,
          holy_symbol: holy_symbol,
          energy: energy,
          preferred_weapon: preferred_weapon,
          devotees: devotees,
          granted_powers: granted_powers,
          obligations_restrictions: obligations_restrictions
        }.compact
      end
    end
  end
end

File: src/ts/models/divindade.ts
import { BaseModel } from "./base.js"
import { Query } from "../query.js"
import { getBookReference, type BookReference } from "./concerns/book_referenceable.js"

const TABLE = "divindades"

/** The type of divine energy a deity channels (affects cleric turning ability). */
export type DivindadeEnergy = "positiva" | "negativa" | "qualquer"

/**
 * A Divindade (deity) in Tormenta20.
 *
 * @example
 * ```ts
 * Divindade.all()
 * Divindade.energiaPositiva().all()
 * Divindade.find("khalmyr")
 *
 * const khalmyr = Divindade.find("khalmyr")!
 * khalmyr.grantedPowers    // ["poder_id", ...]
 * khalmyr.races()          // races that worship Khalmyr
 * khalmyr.classes()        // classes that worship Khalmyr
 * ```
 */
export class Divindade extends BaseModel {
  /** Deity title or epithet (e.g. `"O Senhor da Justiça"`), or `null`. */
  get title(): string | null { return (this._row.title as string) ?? null }
  get description(): string | null { return (this._row.description as string) ?? null }
  /** Beliefs and objectives of the deity's faith. */
  get beliefsObjectives(): unknown[] { return BaseModel.json(this._row.beliefs_objectives, []) }
  get holySymbol(): string | null { return (this._row.holy_symbol as string) ?? null }
  get energy(): DivindadeEnergy { return this._row.energy as DivindadeEnergy }
  /** ID of the deity's preferred weapon, or `null`. */
  get preferredWeapon(): string | null { return (this._row.preferred_weapon as string) ?? null }
  /** Raw devotees object with `races` and `classes` arrays. */
  get devotees(): Record<string, unknown> { return BaseModel.json(this._row.devotees, {}) }
  /** IDs of powers granted to devoted characters. */
  get grantedPowers(): string[] { return BaseModel.json(this._row.granted_powers, []) }
  get obligationsRestrictions(): string | null { return (this._row.obligations_restrictions as string) ?? null }

  /** Returns race IDs whose members commonly worship this deity. */
  races(): string[] { return (this.devotees.races as string[]) ?? [] }
  /** Returns class IDs whose members commonly worship this deity. */
  classes(): string[] { return (this.devotees.classes as string[]) ?? [] }
  /** Returns the sourcebook page reference, or `null` if not indexed. */
  bookReference(): BookReference | null { return getBookReference(this.id) }

  toH() {
    return {
      id: this.id, name: this.name, title: this.title, description: this.description,
      beliefs_objectives: this.beliefsObjectives, holy_symbol: this.holySymbol,
      energy: this.energy, preferred_weapon: this.preferredWeapon,
      devotees: this.devotees, granted_powers: this.grantedPowers,
      obligations_restrictions: this.obligationsRestrictions,
    }
  }

  static query(): Query<Divindade> { return BaseModel.makeQuery(TABLE, Divindade) }
  static all(): Divindade[] { return Divindade.query().all() }
  static find(id: string): Divindade | null { return Divindade.query().find(id) }
  static first(): Divindade | null { return Divindade.query().first() }
  static last(): Divindade | null { return Divindade.query().last() }
  static count(): number { return Divindade.query().count() }

  static byEnergy(energy: DivindadeEnergy): Query<Divindade> { return Divindade.query().where("energy = ?", energy) }
  /** Returns deities that channel positive energy. */
  static energiaPositiva(): Query<Divindade> { return Divindade.byEnergy("positiva") }
  /** Returns deities that channel negative energy. */
  static energiaNegativa(): Query<Divindade> { return Divindade.byEnergy("negativa") }
  /** Returns deities that can channel either energy type. */
  static energiaQualquer(): Query<Divindade> { return Divindade.byEnergy("qualquer") }
}
